### PR TITLE
2th PR - Fetch data from an endpoint

### PR DESCRIPTION
**Pull request by AndresConfortiGL**
## Description
2th PR - Fetch data from an endpoint

## Story Type
- New feature :white_check_mark:
- Change request
- Bug fix
- Tech

## How Can This Be Tested?
By running the app, you will be able to press the button and get prompted by a loading icon, then the data is fetched on the screen, and if an error happens you will get a message instead.

## Evidences
| Pixel 3a | Nexus S |
| ------------- | ------------- |
|pixel|nexus|

| Coverage Test  |
| ------------- |
|coverage|

## Checklist
- All unit Test passed :white_check_mark:

## Tested on
The project was tested on Android 8.0 / API 26 using the following devices:

- Pixel 3a
- Nexus S

## Package Structure
| Structure |
| ------------- |
|Untitled|

### CODE REVIEW

**Review by jnicolaslaurenti — APPROVED** [review]

File: app/src/main/java/com/example/marvelapp/service/request/generator/MarvelRequestGenerator.kt [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    fun <S> createService(serviceClass: Class<S>): S {
+        val retrofit = builder.client(httpClient.build()).build()
+        return retrofit.create(serviceClass)
+    }
+
+    companion object {
+        private const val MARVEL_BASE_URL = "https://gateway.marvel.com"
+        private const val TS = "ts"
+        private const val API = "apikey"
+        private const val HASH = "hash"
```
Comment (DiegoDiazGlobant) [inline app/src/main/java/com/example/marvelapp/service/request/generator/MarvelRequestGenerator.kt:33]: MARVEL_BASE_URL, API_VALUE, and HASH_VALUE should be declared in Gradle as a "buildConfigField"

File: app/src/main/java/com/example/marvelapp/mvvm/viewmodel/CharactersViewModel.kt [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    data class CharactersData(
+        val characterState: CharactersState,
+        val characterInformation: List<Character> = emptyList(),
+        val error: Exception? = null,
+    )
+
+    enum class CharactersState {
+        LOADING,
+        SHOW_CHARACTERS,
+        CHARACTERS_NOT_FOUND,
```
Comment (DiegoDiazGlobant) [inline app/src/main/java/com/example/marvelapp/mvvm/viewmodel/CharactersViewModel.kt]: remove last comma please. It is not necessary

File: app/src/test/java/com/example/marvelapp/MainViewModelTest.kt [hunk 3]
```diff
@@ -0,0 +1,28 @@
+package com.example.marvelapp
```
Comment (DiegoDiazGlobant) [inline app/src/test/java/com/example/marvelapp/MainViewModelTest.kt]: The ViewModel tests should be inside a ViewModel package

File: app/src/main/java/com/example/marvelapp/mapper/MarvelMapper.kt [hunk 4]
```diff
@@ -0,0 +1,20 @@
+package com.example.marvelapp.mapper
+
+import com.example.marvelapp.entity.Character
+import com.example.marvelapp.service.response.DataResponse
+import org.koin.core.component.KoinComponent
+
+class MarvelMapper : KoinComponent{
```
Comment (SartiMau) [inline app/src/main/java/com/example/marvelapp/mapper/MarvelMapper.kt]: KoinComponent is not necessary as you are not injecting anything. In fact, you can remove this and use an extension function

File: app/src/main/java/com/example/marvelapp/mvvm/model/CharactersModel.kt [hunk 5]
```diff
@@ -0,0 +1,10 @@
+package com.example.marvelapp.mvvm.model
+
+import com.example.marvelapp.entity.Character
+import com.example.marvelapp.usecase.GetCharactersUseCase
+import com.example.marvelapp.util.Result
+import org.koin.core.component.KoinComponent
+
+class CharactersModel(private val getCharacterUseCase: GetCharactersUseCase) : KoinComponent {
```
Comment (SartiMau) [inline app/src/main/java/com/example/marvelapp/mvvm/model/CharactersModel.kt]: Extend the KoinComponent here is not necessary

File: app/src/main/java/com/example/marvelapp/mvvm/viewmodel/CharactersViewModel.kt [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+                    )
+                }
+            }
+        }
+    }
+
+    data class CharactersData(
+        val characterState: CharactersState,
+        val characterInformation: List<Character> = emptyList(),
+        val error: Exception? = null,
```
Comment (SartiMau) [inline app/src/main/java/com/example/marvelapp/mvvm/viewmodel/CharactersViewModel.kt]: remove this comma, as it is not necessary

File: app/src/main/java/com/example/marvelapp/mvvm/viewmodel/MainViewModel.kt [hunk 7]
```diff
@@ -1,6 +1,22 @@
...
+    fun onPressedButton() {
+        _buttonState.value = ButtonData(ButtonState.PRESSED_BUTTON)
+    }
+
+    data class ButtonData(
+        val buttonPressed: ButtonState
+    )
+
+    enum class ButtonState {
+        PRESSED_BUTTON,
```
Comment (SartiMau) [inline app/src/main/java/com/example/marvelapp/mvvm/viewmodel/MainViewModel.kt]: same

File: app/src/main/java/com/example/marvelapp/service/CharacterServiceImpl.kt [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+                    return Result.Success(it)
+                }
+        } catch (e: Exception) {
+            return Result.Failure(Exception(FAILED_CHARACTERS_GET))
+        }
+        return Result.Failure(Exception(FAILED_CHARACTERS_GET))
+    }
+
+    companion object {
+        private const val FAILED_CHARACTERS_GET = "COULDN'T FIND ANY CHARACTERS"
```
Comment (SartiMau) [inline app/src/main/java/com/example/marvelapp/service/CharacterServiceImpl.kt]: if you will show this message on the screen, this has to be in a String resource

**Review by SartiMau — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/example/marvelapp/service/request/generator/MarvelRequestGenerator.kt (hunk 1 above)
Comment (AndresConfortiGL) [reply app/src/main/java/com/example/marvelapp/service/request/generator/MarvelRequestGenerator.kt:33]: fixed

File: app/src/main/java/com/example/marvelapp/mvvm/viewmodel/CharactersViewModel.kt (hunk 2 above)
Comment (AndresConfortiGL) [reply app/src/main/java/com/example/marvelapp/mvvm/viewmodel/CharactersViewModel.kt]: fixed

File: app/src/test/java/com/example/marvelapp/MainViewModelTest.kt (hunk 3 above)
Comment (AndresConfortiGL) [reply app/src/test/java/com/example/marvelapp/MainViewModelTest.kt]: fixed

File: app/src/main/java/com/example/marvelapp/mvvm/model/CharactersModel.kt (hunk 5 above)
Comment (AndresConfortiGL) [reply app/src/main/java/com/example/marvelapp/mvvm/model/CharactersModel.kt]: fixed

File: app/src/main/java/com/example/marvelapp/mvvm/viewmodel/CharactersViewModel.kt (hunk 6 above)
Comment (AndresConfortiGL) [reply app/src/main/java/com/example/marvelapp/mvvm/viewmodel/CharactersViewModel.kt]: fixed

File: app/src/main/java/com/example/marvelapp/mvvm/viewmodel/MainViewModel.kt (hunk 7 above)
Comment (AndresConfortiGL) [reply app/src/main/java/com/example/marvelapp/mvvm/viewmodel/MainViewModel.kt]: fixed

File: app/src/main/java/com/example/marvelapp/service/CharacterServiceImpl.kt (hunk 8 above)
Comment (AndresConfortiGL) [reply app/src/main/java/com/example/marvelapp/service/CharacterServiceImpl.kt:30]: fixed

File: app/src/main/java/com/example/marvelapp/activity/CharactersActivity.kt [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+        setContentView(binding.root)
+        viewModel.getCharacters()
+        viewModel.characterState.observe({ lifecycle }, ::charactersState)
+    }
+
+    private fun charactersState(CharacterData: CharactersViewModel.CharactersData) {
+        when (CharacterData.characterState) {
+            CharactersViewModel.CharactersState.LOADING -> displayLoader()
+            CharactersViewModel.CharactersState.SHOW_CHARACTERS -> displayCharacters(CharacterData.characterInformation)
+            CharactersViewModel.CharactersState.CHARACTERS_NOT_FOUND -> displayError(R.string.main_activity_get_error)
```
Comment (mariano98montero) [inline app/src/main/java/com/example/marvelapp/activity/CharactersActivity.kt:31]: why you send the exception if you are not going to display any information about it?

File: app/src/main/java/com/example/marvelapp/service/CharacterServiceImpl.kt (hunk 8 above)
Comment (AndresConfortiGL) [reply app/src/main/java/com/example/marvelapp/service/CharacterServiceImpl.kt]: fixed

File: app/src/main/java/com/example/marvelapp/mapper/MarvelMapper.kt (hunk 4 above)
Comment (AndresConfortiGL) [reply app/src/main/java/com/example/marvelapp/mapper/MarvelMapper.kt]: fixed

**Review by mariano98montero — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/example/marvelapp/activity/CharactersActivity.kt (hunk 9 above)
Comment (AndresConfortiGL) [reply app/src/main/java/com/example/marvelapp/activity/CharactersActivity.kt:31]: fixed

**Review by SartiMau — APPROVED** [review]

**Review by Georgina-Fernandez — APPROVED** [review]

**Review by mariano98montero — APPROVED** [review]

**Pull request merged**